validator/client/beacon-api/get_beacon_block.go

```
@@ -22,8 +25,28 @@ func (c *beaconApiValidatorClient) beaconBlock(ctx context.Context, slot primiti
 	if len(graffiti) > 0 {
 		queryParams.Add("graffiti", hexutil.Encode(graffiti))
 	}
-
 	queryUrl := apiutil.BuildURL(fmt.Sprintf("/eth/v3/validator/blocks/%d", slot), queryParams)
+	if features.Get().EnableSSZ {
+		data, header, err := c.jsonRestHandler.GetSSZ(ctx, queryUrl)
+		if err != nil {
+			return nil, err
```

should i fall back to JSON?

---

I think it would be good if we could distinguish between a ssz-not-supported error and other errors. in that way we could either return the error or try json. 

I also think the validator client should remember which endpoints of the beacon client support ssz and which don't. that way it won't get errors more than once. 
this could be cached with a TTL so changes to the beacon node can go into effect.

---

in terms of it remembering why wouldn't it just have this feature flag check because internally we may need to handle it differently, it'll be the most flexible. maybe i'm not fully understanding what you intend to do with caching there's no meta endpoint that provides the information so I think we just manually do it. we can propose a meta endpoint to beacon apis if we want to describe the versions and types it supports if we want it smarter but probably overkill. 

typically the ssz not supported would mean a version mismatch or something, the beacon api should return a 415 error if truly not supported we can handle that and fall back to json if we want

---

yes, but if the ssz flag is used, and we call a beacon api every slot, that does not support ssz, we have to go through the process of getting the 415 error, and falling back to json everytime. which could be avoided if we cache that endpoint X does not support ssz, so don't even try it.

---

I see that's fair, i don't know if I want to include that in this pr, it sounds like a nice to have ( to me) maybe some others will chime in

---

we definitely don't have this for builder APIs either, if it doesn't support it'll just fail ( no one complained yet in issues)